### PR TITLE
A buyer is now to pay via stripe

### LINKED ISSUE #81: Feat: Secure Payment Processing with Stripe API
**Story Summary:**

As a buyer, I want to securely complete my purchase by processing the payment through the Stripe API so that I can receive confirmation of my order and ensure a smooth transaction.

**Acceptance Criteria**

 - GIVEN a valid payment request is sent to the backend
- WHEN the request is authenticated and the request body is valid
### THEN:
- [ ] The payment should be successfully processed using the Stripe API
- [ ] The order status should be updated to "Paid" in the database
- [ ] A confirmation response should be sent to the frontend, including the order number, total cost, and expected delivery date
WHEN the request body is invalid or authentication fails
THEN:
- [ ] An appropriate error response should be sent to the frontend, describing the validation error

**Technical Requirements:**

- [ ] All inputs should be validated for sanity and security
- [ ] The Stripe API key should be kept secure and not exposed in the backend code
- [ ] The Stripe library should be used to make the payment, charge the customer, and verify the transaction
- [ ] The order status should be updated in real-time in both the database and the frontend to reflect the payment status.

### PR DESCRIPTION
**Pull request by EddyShimwa**
### Title: Secure Payment Processing with Stripe API

**Description:**

This PR introduces functionality to securely process payments through the Stripe API. Upon successful payment, the `paid` field in the database will be set to `True`. It also handles cases where the payment request is invalid or authentication fails.

**Related User Story:** As a buyer, I want to securely complete my purchase by processing the payment through the Stripe API so that I can receive confirmation of my order and ensure a smooth transaction.

**Acceptance Criteria:**

*   **GIVEN** a valid payment request is sent to the backend
*   **WHEN** the request is authenticated and the request body is valid
    *   [ ] The payment should be successfully processed using the Stripe API
    *   [ ] The order status should be updated to "Paid" in the database
    *   [ ] A confirmation response should be sent to the frontend, including the order number, total cost, and expected delivery date
*   **WHEN** the request body is invalid or authentication fails
    *   [ ] An appropriate error response should be sent to the frontend, describing the validation error
**Technical Requirements:**

*   \[ \] All inputs should be validated for sanity and security
*   \[ \] The Stripe API key should be kept secure and not exposed in the backend code
*   \[ \] The Stripe library should be used to make the payment, charge the customer, and verify the transaction
*   \[ \] The order status should be updated in real-time in both the database and the frontend to reflect the payment status

**Changes Introduced:**

1.  **Payment Processing:**
    
    *   Integrated Stripe API to handle payment processing.
    *   Created service to interact with Stripe and handle payment logic.
2.  **Order Status Update:**
    
    *   Updated order status to "Paid" in the database upon successful payment.
    *   Ensured real-time updates to the frontend regarding payment status.
3.  **Response Handling:**
    
    *   Sent confirmation response to the frontend including order number, total cost, and expected delivery date.
    *   Handled error responses for invalid requests or authentication failures.
4.  **Security and Validation:**
    
    *   Validated all inputs for sanity and security.
    *   Ensured the Stripe API key is securely managed and not exposed in the codebase.

**Files Modified:**

*   `controllers/buyerController.ts`
*   `docs/buyerDocs.ts`
*   `models/orderEntity.ts`
*   `routes/buyerRoutes`

**How to Test:**

1.  **Valid Payment Request:**
    
    *   Send a valid payment request to the backend.
    *   Verify that the payment is processed successfully.
    *   Check that Paid field for that order is set to true in the database.

2.  **Invalid Payment Request:**
    *   Send an invalid payment request to the backend.
    *   Verify that an appropriate error response is sent to the frontend.
3.  **Security Checks:**
    
    *   Ensure that no sensitive information, such as the Stripe API key, is exposed in the codebase.
    *   Validate that all inputs are properly sanitized and secured.

**Screenshots:**
screen1
screen2
screen3

**Additional Notes:**

-   Ensure the Stripe API key is set in the environment variables before testing.
-  Run the tests using `npm run test` to validate the new changes.

**Reviewer Checklist:**

 - [ ] All acceptance criteria are met.
 - [ ]  Code is clean, well-documented, and follows best practices.
 - [ ]  No sensitive information is exposed.
 - [ ] Tests have been added and/or updated to cover the new functionality.
 - [ ] The application runs without errors and all functionalities are verified.

* * *

Please review the changes and let me know if there are any issues or further improvements needed. Thank you!

#81

### CODE REVIEW

File: src/__mocks__/mailer.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    sendCode: jest.fn((to, subject, htmlTemplatePath, replacements) => {
+      console.log('Mock sendCode called with:', { to, subject, htmlTemplatePath, replacements });
```
Comment (wayneleon1) [inline src/__mocks__/mailer.ts]: try to remove all those console

**Review by ambroisegithub — DISMISSED** [review]
It Looks good for merge!!!

**Review by jkarenzi — DISMISSED** [review]

**Review by jkarenzi — DISMISSED** [review]

**Review by ambroisegithub — DISMISSED** [review]
Great job!!!

**Review by wayneleon1 — DISMISSED** [review]
great job

**Review by wayneleon1 — APPROVED** [review]
Looks good

**Review by ambroisegithub — DISMISSED** [review]
Great Job!!!

**Review by Dawaic6 — DISMISSED** [review]

**Review by wayneleon1 — DISMISSED** [review]
Good Job

File: src/__mocks__/mailer.ts (hunk 1 above)
Comment (dushimeemma) [inline src/__mocks__/mailer.ts]: you can remove the console log statement

**Review by jkarenzi — DISMISSED** [review]

File: src/__test__/buyerController.tesst.ts
Comment (dushimeemma) [inline src/__test__/buyerController.tesst.ts:1]: you can rename the file with the extension `.test.ts` unless you have any special reason why

File: src/__test__/category.tesst.ts
Comment (dushimeemma) [inline src/__test__/category.tesst.ts:1]: you can rename the file with extension `.test.ts` unless you have any special reason why

File: src/__test__/changeStatus.tesst.ts
Comment (dushimeemma) [inline src/__test__/changeStatus.tesst.ts:1]: file renaming as well 👆

File: src/__test__/product.tesst.ts
Comment (dushimeemma) [inline src/__test__/product.tesst.ts:1]: work file renaming 👆

File: src/__test__/roles.tesst.ts
Comment (dushimeemma) [inline src/__test__/roles.tesst.ts:1]: file renaming 👆

Comment (EddyShimwa) [reply src/__test__/roles.tesst.ts:1]: I just added those files, when i was trying to solve something, but i realised that i forgot to deleted them! So i deleted them!

**Review by wayneleon1 — DISMISSED** [review]
Good Job

File: src/__test__/roles.tesst.ts
Comment (dushimeemma) [inline src/__test__/roles.tesst.ts:1]: you forgot to refactor this one 👆

**Review by dushimeemma — APPROVED** [review]
Good job 🥇 🥇 🥇

**Review by Dawaic6 — APPROVED** [review]

**Review by bertrandshema — APPROVED** [review]
All good now

**Review by jkarenzi — APPROVED** [review]

**Pull request merged**